common/crypto/platform/openssl/crypto.cpp

```diff
@@ -430,6 +439,97 @@ expected::ExpectedString SignRawData(
 	return Sign(private_key_path, shasum);
 }
 
+const size_t ecdsa256curveBits = 256;
+const size_t ecdsa256keySize = 32;
+
+// Try and decode the keys from pure binary, assuming that the points on the
+// curve (r,s), have been concatenated together (r || s), and simply dumped to
+// binary. Which is what we did in the `mender-artifact` tool.
+// (See MEN-1740) for some insight into previous issues, and the chosen fix.
+template <BIGNUM *DecoderFn(const unsigned char *, int, BIGNUM *)>
+static expected::ExpectedBytes TryASN1EncodeMenderCustomBinaryECFormat(
+	const vector<uint8_t> &signature, const mender::sha::SHA &shasum) {
+	// Verify that the marshalled keys match our expectation
+	const size_t assumed_signature_size {2 * ecdsa256keySize};
+	if (signature.size() > assumed_signature_size) {
+		return expected::unexpected(MakeError(
+			SetupError,
+			"Unexpected size of the signature for ECDSA. Expected 2*32 bytes. Got: "
+				+ to_string(signature.size())));
+	}
+	auto ecSig = unique_ptr<ECDSA_SIG, void (*)(ECDSA_SIG *)>(ECDSA_SIG_new(), ECDSA_SIG_free);
+	if (ecSig == nullptr) {
+		return expected::unexpected(MakeError(
+			SetupError,
+			"Failed to allocate the structure for the ECDSA signature" + GetOpenSSLErrorMessage()));
+	}
+
+	auto r = unique_ptr<BIGNUM, void (*)(BIGNUM *)>(
+		DecoderFn(signature.data(), ecdsa256keySize, NULL), BN_free);
+	if (r == NULL) {
+		return expected::unexpected(MakeError(
+			SetupError,
+			"Failed to extract the r(andom) part from the ECDSA signature in the binary representation"
+				+ GetOpenSSLErrorMessage()));
+	}
+	auto s = unique_ptr<BIGNUM, void (*)(BIGNUM *)>(
+		DecoderFn(signature.data() + ecdsa256keySize, ecdsa256keySize, NULL), BN_free);
+	if (s == NULL) {
+		return expected::unexpected(MakeError(
+			SetupError,
+			"Failed to extract the s(ignature) part from the ECDSA signature in the binary representation"
+				+ GetOpenSSLErrorMessage()));
+	}
+
+	// Set the r&s values in the SIG struct
+	// r & s now owned by ecSig
+	int ret {ECDSA_SIG_set0(ecSig.get(), r.release(), s.release())};
```

Comment:
Possible subtle gotcha here: If `ECDSA_SIG_set0` fails, I assume that it will not take ownership. So you should actually use `.get()` here, and then do `.release()` after the check below instead. Thanks to @lluiscampos for making me aware of this in an earlier PR.

Reply:
Uhhh, very good point! :grin:

Reply:
> I assume that it will not take ownership.

I was trying to check this in the man page, but it doesn't mention what the exact behavior is. Not sure if a potential leak here is not better than a potential double-free.

Reply:
The `manpage` does this in the example:
```
     if (!ECDSA_SIG_set0(obj, rbn, sbn))
            /* error */
        /* Set these to NULL since they are now owned by obj */
        rbn = sbn = NULL;
```

Reply:
So I guess releasing after is the way to go

Reply:
Yes, that's what I think too